From 18b81309729a9d0dda61aaebba3dd4f06d63d8ee Mon Sep 17 00:00:00 2001
From: K <kayo@illumium.org>
Date: Mon, 27 Feb 2023 17:32:28 +0500
Subject: [PATCH 113/114] pwm-sunxi-enhance: Port to linux 6.2.y

---
 drivers/pwm/pwm-sunxi-enhance.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/pwm/pwm-sunxi-enhance.c b/drivers/pwm/pwm-sunxi-enhance.c
index 0c9d36cdd47e..11c48149393c 100644
--- a/drivers/pwm/pwm-sunxi-enhance.c
+++ b/drivers/pwm/pwm-sunxi-enhance.c
@@ -941,7 +941,7 @@ static int sunxi_pwm_apply(struct pwm_chip *chip, struct pwm_device *pwm,
 	return err;
 }
 
-static void sunxi_pwm_get_state(struct pwm_chip *chip, struct pwm_device *pwm,
+static int sunxi_pwm_get_state(struct pwm_chip *chip, struct pwm_device *pwm,
 				struct pwm_state *state)
 {
 	struct sunxi_pwm_chip *pc = to_sunxi_pwm_chip(chip);
@@ -951,6 +951,8 @@ static void sunxi_pwm_get_state(struct pwm_chip *chip, struct pwm_device *pwm,
 	state->duty_cycle = 0;
 	sunxi_pwm_get_polarity(chip, pwm, &state->polarity);
 	state->enabled = sunxi_pwm_enabled(chip, pwm);
+
+	return 0;
 }
 
 static struct pwm_ops sunxi_pwm_ops = {
-- 
2.38.3

