From 033498c5e3bed888011d26afbaf7640b98516044 Mon Sep 17 00:00:00 2001
From: K <kayo@illumium.org>
Date: Mon, 6 Dec 2021 10:49:41 +0500
Subject: [PATCH 1/3] Adding h616 THS workaround.

---
 board/sunxi/board.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/board/sunxi/board.c b/board/sunxi/board.c
index 21a2407e06..95fcce42c0 100644
--- a/board/sunxi/board.c
+++ b/board/sunxi/board.c
@@ -251,6 +251,15 @@ int board_init(void)
 		}
 	}
 
+#if CONFIG_MACH_SUN50I_H616
+	/*
+	 * The bit[16] of register reg[0x03000000] must be zero for the THS
+	 * driver to work properly in the kernel. The BSP u-boot is putting
+	 * the whole register to zero so we are doing the same.
+	 */
+	writel(0x0, SUNXI_SRAMC_BASE);
+#endif
+
 #if CONFIG_IS_ENABLED(DM_I2C)
 	/*
 	 * Temporary workaround for enabling I2C clocks until proper sunxi DM
-- 
2.36.0

